src/witness_generator/generator.rs

```diff
@@ -305,13 +305,21 @@ where
         // Note that we should always query all machines that match, because they might
         // update their internal data, even if all values are already known.
         // TODO could it be that multiple machines match?
+        let mut fixed_lookup = None;
         for m in &mut self.machines {
             // TODO also consider the reasons above.
-            if let Some(result) =
-                m.process_plookup(self.fixed_data, identity.kind, &left, &identity.right)
-            {
+            if let Some(result) = m.process_plookup(
+                self.fixed_data,
+                &mut fixed_lookup,
```

Comment:
This is a bit weird. I want to pass a reference to the fixed lookup machine, i.e. the first in the machines.
It has to be a mutable reference because the fixed lookup updates its cache.
It has to be an option, because the fixed lookup cannot have a mutable reference to itself in the call to `process_plookup`.
Just using `Option<&mut dyn Machine>` didn't work, because then I would move `fixed_lookup`. This is the only way I could make this work, but I wonder if the double mut ref is really necessary.

Reply:
I don't have a suggestion for exactly this but I'm not sure I'm convinced by treating the `FixedLookup` like a normal machine:
- we expect a specific implementation of Machine in `process_lookup` for the `fixed_lookup` argument
- this specific piece of code seems to rely on ordering in the machine vector to not pass any `FixedLookup` to the first machine, which is assumed to be `FixedLookup`?
- Don't we always have a single `FixedLookup` ?

I understand that putting everything under the `Machine` trait creates some encapsulation, but it seems like some code becomes simpler if we don't do that.

I had a quick go at extracting it [here](https://github.com/chriseth/powdr/compare/block_machines...separate-fixed-lookup?expand=1) 

Am I missing something on why `FixedLookup` needs to be abstracted away as a `Machine`?

Reply:
Yeah, that works as well. I just thought it fits the interface and we don't need to "unroll" the loop. But yeah, the complication with the `&mut Option<&mut dyn Machine>` is probably not worth it.